### PR TITLE
[mojo-stdlib] Refactor tests to use `assert_*` functions whenever possible

### PR DESCRIPTION
**Pull request by shenyih0ng**
# Description

This PR ports a subset of tests using `FileCheck` to use `testing.assert_*` functions (#2024)

Below is a list of files still using `FileCheck` (`git grep -l '| FileCheck'`) as of 8bf9e6c5b4693f2d666c14f2260f11a33f7fb955, and the highlighted files are the ones I've tackled:

```diff
examples/deviceinfo.mojo
"examples/hello.\360\237\224\245"
examples/mandelbrot.mojo
examples/reduce.mojo
stdlib/docs/development.md
- stdlib/test/base64/test_base64.mojo
stdlib/test/builtin/test_bfloat16.mojo
stdlib/test/builtin/test_debug_assert-error.mojo
stdlib/test/builtin/test_debug_assert-mojo-enable-assertions.mojo
stdlib/test/builtin/test_debug_assert.mojo
stdlib/test/builtin/test_debug_assert_warning.mojo
- stdlib/test/builtin/test_error.mojo
stdlib/test/builtin/test_file.mojo
- stdlib/test/builtin/test_int_literal.mojo
- stdlib/test/builtin/test_issue_1004.mojo
- stdlib/test/builtin/test_list.mojo
stdlib/test/builtin/test_object.mojo
stdlib/test/builtin/test_print.mojo
stdlib/test/builtin/test_print_long_string.mojo
- stdlib/test/builtin/test_range.mojo
stdlib/test/builtin/test_simd.mojo
stdlib/test/builtin/test_slice.mojo
stdlib/test/memory/test_anypointer.mojo
stdlib/test/memory/test_bitcast.mojo
- stdlib/test/os/test_atomic.mojo
stdlib/test/os/test_no_trap.mojo
stdlib/test/os/test_trap.mojo
stdlib/test/os/test_trap_stringable.mojo
- stdlib/test/python/test_ownership.mojo
- stdlib/test/python/test_python_error_handling.mojo
- stdlib/test/python/test_python_info.mojo
- stdlib/test/python/test_python_interop.mojo
- stdlib/test/python/test_python_object.mojo
- stdlib/test/python/test_python_to_mojo.mojo
stdlib/test/random/test_random.mojo
stdlib/test/sys/test_aarch64_target.mojo
stdlib/test/sys/test_build_info_debug.mojo
stdlib/test/sys/test_has_intel_amx.mojo
- stdlib/test/sys/test_intrinsics.mojo
stdlib/test/sys/test_linux_target.mojo
- stdlib/test/sys/test_macos_target.mojo
stdlib/test/sys/test_paramenv.mojo
- stdlib/test/sys/test_targetinfo.mojo
stdlib/test/sys/test_windows_target.mojo
stdlib/test/testing/test_assert_raises_basic.mojo
stdlib/test/testing/test_assert_raises_no_error.mojo
stdlib/test/testing/test_assert_raises_no_match.mojo
- stdlib/test/time/test_time.mojo
- stdlib/test/utils/issue_13632.mojo
- stdlib/test/utils/test_numerics.mojo
stdlib/test/utils/test_unroll.mojo
```

### CODE REVIEW

File: stdlib/test/python/test_python_to_mojo.mojo
Comment (peymanbr) [inline stdlib/test/python/test_python_to_mojo.mojo:1]: FileCheck has to be removed.

File: stdlib/test/sys/test_intrinsics.mojo
Comment (peymanbr) [inline stdlib/test/sys/test_intrinsics.mojo:1]: FileCheck in the line #13 has to be removed.

File: stdlib/test/python/test_python_to_mojo.mojo
Comment (shenyih0ng) [reply stdlib/test/python/test_python_to_mojo.mojo:1]: `FileCheck` is still required here as I am relying it to check for `range` correctness

**Review by JoeLoser — COMMENTED** [review]
This is great, thank you very much for your contribution in reworking many of the tests! 🎉

**Review by JoeLoser — APPROVED** [review]
Looks great! You'll just need to rebase and resolve the one conflict in `stdlib/test/os/test_atomic.mojo`, and then I'm happy to merge this!

**Comment by shenyih0ng** [comment]
> Looks great! You'll just need to rebase and resolve the one conflict in `stdlib/test/os/test_atomic.mojo`, and then I'm happy to merge this!

Thanks for the review! I have rebased and resolved the conflict

**Comment by JoeLoser** [comment]
> > Looks great! You'll just need to rebase and resolve the one conflict in `stdlib/test/os/test_atomic.mojo`, and then I'm happy to merge this!
> 
> Thanks for the review! I have rebased and resolved the conflict

Merged!  Thanks again for the contribution 🎉!

**Pull request merged**